### DISCUSSION TITLE
models/fast-sam/

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#15914] [Docs]
# models/fast-sam/

Discover FastSAM, a real-time CNN-based solution for segmenting any object in an image. Efficient, competitive, and ideal for various vision tasks.

https://docs.ultralytics.com/models/fast-sam/

**Comment by YuechengLiu** [comment]
Thanks the authors for the great work!

I'm using FastSAM-x for instance segmentation (bboxes as the prompt), but I find that the shape of the predicted masks mismatch with the bounding box number. (e.g. the model takes 18 bboxes as input, but the shape of the predicted mask is (17, H, W)) . I tried to set `conf=0.01` and `iou=0.999`, but it doesn't work.  How can I solve that? Thanks!

**Reply by glenn-jocher** [reply]
@YuechengLiu thank you for your kind words! The issue you're experiencing might be due to a mismatch in the bounding box prompts and the segmentation results. Please ensure you're using the latest version of the FastSAM model. If the problem persists, consider checking the input format and the bounding box coordinates. For further assistance, you can refer to the FastSAM documentation: https://docs.ultralytics.com/models/fast-sam/.

**Reply by YuechengLiu** [reply]
Thanks for your kind reply! The bounding boxes I'm using are predicted by yolov10x (result.boxes.xyxy), so I think the input format might be right. I also tried the same bounding boxes for `mobile_sam`, and the output dimension is as expected (same as the input dimension). The problem does not always occur. Do you have any idea about that? Thanks!

**Reply by YuechengLiu** [reply]
And I'm using the latest version of ultralytics

**Reply by glenn-jocher** [reply]
Thanks for confirming you're using the latest version. Given that the issue is intermittent, it might be related to the specific bounding boxes or the image content. Please try running the model with a different set of bounding boxes or images to see if the issue persists. If it does, consider opening an issue on the GitHub repository with detailed information for further investigation.

**Comment by feiwenxiong** [comment]
how many boxes are supported in thsi api, it cound only segment 2 boxes at most.

**Reply by glenn-jocher** [reply]
@feiwenxiong fastSAM should support multiple boxes. If you're experiencing limitations, please ensure your input format is correct. For further assistance, you might want to check the latest updates or documentation.

**Comment by chongkuiqi** [comment]
Thanks for your great work！ 
I'm using FastSAM-x with bboxes prompt, the original image shape is [720, 1280], and I set 'imgsz' 1024. 
However, the results mask shape is [576, 1024], but I remembered in ultralytics-8.2.10, I can get masks with the same shape as the original image. May I ask why this change happened ?

**Reply by chongkuiqi** [reply]
Sorry, I forgot to set 'Retina_masks', problem solved, thanks again.

**Reply by glenn-jocher** [reply]
Glad to hear the issue is resolved! If you have any more questions, feel free to ask.

**Comment by ismailwale** [comment]
How can I  make an inference of a video using frame by frame in loop using FastSam ?because I tried multiple times making inference directly with video predictor I constantly running out of memory despite setting stream =True since SAM is trying to keep the memory of 2000 frames of the video at ones . I will appreciate any ideal on how to predict and segment the video as a single frame image at every instance and  put everything in a loop.

Similar question win MobileSam

**Reply by glenn-jocher** [reply]
@ismailwale to process a video frame by frame and avoid memory overload with FastSAM or MobileSAM, load the video as a frame sequence using a tool like OpenCV. Loop through each frame, run inference on it individually, and clear any cached results to free up memory between frames. For reference on frame-wise usage, consult OpenCV and FastSAM integration. Let me know if you need further clarifications!

**Reply by ismailwale** [reply]
Thanks so much for your guide .I appreciate it sir.

**Reply by glenn-jocher** [reply]
Thank you for your appreciation! I'm glad the guide was useful. Please feel free to reach out if you have further questions about utilizing FastSAM or MobileSAM effectively for your tasks.

**Reply by llu025** [reply]
Thank you all for your great work! I am having troubles with prompting with a bounding box, it seems the FastSAM model ignores it and detects objects all over the video frames. This is how I obtain the results:
```
results = model(
    input_filename,
    device=device,
    imgsz=(736, 1280),
    verbose=False,
    stream=True,
    bboxes=[0, 0, 360, 640],
    retina_masks=True,
    conf=0.4,
    iou=0.9,
)
```

I also tried with `bboxes=[[0, 0, 360, 640]],` without success. The model should detect objects in the top left quadrant of the video, instead it detects everywhere. Also, it oddly does not raise any error or warnings, both with `bboxes` as list of integers or `bboxes` as list of lists of integers.

**Reply by glenn-jocher** [reply]
It seems the bounding box prompts aren't being applied as expected. Ensure the `bboxes` format is correct (e.g., `[[x1, y1, x2, y2]]` as a list of lists). Additionally, try using the `FastSAMPredictor` approach for more controlled prompt application, as shown in [this example](https://docs.ultralytics.com/models/fast-sam/#prompt-inference). Let me know if the issue persists!

**Reply by llu025** [reply]
Thank you @glenn-jocher for the "prompt" response ;-)

I need to use `stream=True` since I am working with large videos. I cannot use it coupled with `FastSAMPredictor`, because it does not support it. My only alternative is to grab frames on my own with OpenCV, is that correct?

By the way, in your examples in [here](https://docs.ultralytics.com/models/fast-sam/#predict-usage) `bboxes` is passed as a list of integers, I followed those in the beginning. I see in the examples with `FastSAMPredictor` that a list of lists is used instead. In any case, I ensured that I was using a list of lists, and it did not work anyway.

Here's a minimal example of what I am trying to do, i.e., detecting objects in the top left part of the video:

```
from ultralytics import FastSAM
import os
import ffmpeg

input_filename = "test_video.mp4"
base_name = os.path.splitext(os.path.basename(input_filename))[0]
output_filename = os.path.join("output", f"{base_name}.avi")

model = FastSAM("models/FastSAM-x.pt")

process2 = (
    ffmpeg.input(
        "pipe:",
        format="rawvideo",
        pix_fmt="rgb24",
        s="1280x720",
        loglevel="-8",
    )
    .output(
        output_filename,
        pix_fmt="yuv420p",
        vcodec="libx264",
        loglevel="-8",
        crf=17,
    )
    .overwrite_output()
    .run_async(pipe_stdin=True)
)

results = model(
    input_filename,
    device="cuda",
    imgsz=(736, 1280),
    verbose=False,
    stream=True,
    bboxes=[[0, 0, 360, 640]],
    retina_masks=True,
    conf=0.4,
    iou=0.9,
)

for result in results:
    frame = result.plot()
    # Save on disk with process2
    process2.stdin.write(frame.tobytes())

process2.stdin.close()
process2.wait()
```

**Reply by glenn-jocher** [reply]
Thank you for the detailed explanation and example. Yes, you're correct—using OpenCV to manually grab frames is a valid and efficient approach, especially when working with `stream=True` for large videos, as it avoids memory issues. Regarding `bboxes`, FastSAM expects a list of lists in the format `[[x1, y1, x2, y2]]`. If this still doesn't work as intended, please ensure your bounding box coordinates are correct and within the frame dimensions. Let us know if you encounter any further issues!

**Reply by llu025** [reply]
I still encounter the same issue. I used `bboxes=[[20, 20, 500, 500]]` just to be 100% sure the coordinates are within the the frame dimensions (1280x720), nothing changed. Since I have to grab frames with OpenCV anyway, I will just crop them myself, but I am quite sure this option is not working as intended.

**Reply by glenn-jocher** [reply]
@llu025 thank you for the update. If you’re certain the `bboxes` parameter is properly configured and still not working as intended, cropping frames manually with OpenCV before passing them to the model is indeed a reliable workaround. It may also be worth checking the latest FastSAM updates or reporting this as a potential issue to ensure the bounding box prompt functionality is reviewed. Let us know if you need further assistance!

**Comment by shahedmomenzadeh** [comment]
Hi, I have a question
Can I use the generated bbox cordinates by a yolo detection model like yolov11 as bboxes prompt for FastSAM ?

**Reply by glenn-jocher** [reply]
@shahedmomenzadeh yes, you can use YOLO detection model-generated bbox coordinates as prompts for FastSAM. The [FastSAM documentation](https://docs.ultralytics.com/models/fast-sam/#predict-usage) shows examples of passing bbox coordinates directly to the model using either the Python API or CLI. Just ensure your coordinates are in XYXY format as shown in the examples.
